packages/react-import-remote/src/hooks.ts

```diff
@@ -0,0 +1,100 @@
+import * as React from 'react';
+import {useIntersection} from '@shopify/react-intersection-observer';
+import {
+  DeferTiming,
+  WindowWithRequestIdleCallback,
+  RequestIdleCallbackHandle,
+} from '@shopify/async';
+import load from './load';
+
+interface Options<Imported> {
+  nonce?: string;
+  defer?: DeferTiming;
+  onError?(error: Error): void;
+  getImport(window: Window): Imported;
+  onImported?(imported: Imported): void;
+}
+
+type State = 'loading' | 'error' | 'loaded' | null;
+
+export function useImportRemote<Imported = any>(
+  source: string,
+  options: Options<Imported>,
+): {
+  state: State;
+  imported: Imported | null;
+  intersectionRef: any;
+} {
+  const [intersection, intersectionRef] = useIntersection();
+  const [state, setState] = React.useState<State>(null);
+  const [imported, setImported] = React.useState<Imported | null>(null);
+  const idleCallbackHandle = React.useRef<RequestIdleCallbackHandle | null>(
+    null,
+  );
+  const {
+    defer = DeferTiming.Mount,
+    nonce = '',
+    getImport,
+    onError = () => {},
+    onImported = () => {},
+  } = options;
+
+  const loadRemote = React.useCallback(
+    () => {
+      return new Promise(async resolve => {
+        try {
+          setState('loading');
+          const importResult = await load(source, getImport, nonce);
+          setImported(importResult);
+          onImported(importResult);
```

Comment:
I don't think this callback really makes sense in a hook, personally (it kind of makes sense in the component version, but you can implement it using `useOnValueChange` there instead